Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go

```diff
@@ -0,0 +1,365 @@
+					sent_need_pods = true
+					timeout_started = true
+				}
+
+				if allConns.Len() == 0 {
+					if !proxier.loadBalancer.ServiceHasEndpoints(service) {
+						// notify us when endpoints are available
+						go tcp.waitForEndpoints(endpointsAvail, service, proxier)
+					}
+				}
+
+				allConns.Add(inConn)
+				glog.V(4).Infof("unidling TCP proxy has accumulated %v connections while waiting for service %s/%s:%s to unidle", allConns.Len(), service.Namespace, service.Name, service.Port)
+			case <-ticker.C:
+				if !timeout_started {
+					continue WAIT_LOOP
+				}
+				// TODO: timeout each connection (or group of connections) separately
+				// timed out, close all waiting connections and reset the state
+				allConns.Tick()
+			}
+		}
+
+		ticker.Stop()
+	}
+
+	glog.V(4).Infof("unidling TCP proxy waiting for endpoints for service %s/%s:%s to become available with %v accumulated connections", service.Namespace, service.Name, service.Port, allConns.Len())
+	// block until we have endpoints available
+	select {
+	case _, ok := <-endpointsAvail:
+		if ok {
+			close(endpointsAvail)
+			// this shouldn't happen (ok should always be false)
+		}
+	case <-time.NewTimer(needPodsWaitTimeout).C:
+		if allConns.Len() > 0 {
+			glog.Errorf("timed out %v TCP connections while waiting for idled service %s/%s:%s to awaken.", allConns.Len(), service.Namespace, service.Name, service.Port)
+			allConns.Clear()
+		}
+		return
+	}
+	glog.V(4).Infof("unidling TCP proxy got endpoints for service %s/%s:%s, connecting %v accumulated connections", service.Namespace, service.Name, service.Port, allConns.Len())
+
+	for _, inConn := range allConns.GetConns() {
+		glog.V(3).Infof("Accepted TCP connection from %v to %v", inConn.RemoteAddr(), inConn.LocalAddr())
+		outConn, err := tryConnect(service, inConn.(*net.TCPConn).RemoteAddr(), "tcp", proxier)
+		if err != nil {
+			glog.Errorf("Failed to connect to balancer: %v", err)
+			inConn.Close()
+			continue
+		}
+		// Spin up an async copy loop.
+		go proxyTCP(inConn.(*net.TCPConn), outConn.(*net.TCPConn))
+	}
+}
+
+// udpUnidlerSocket implements proxySocket.  Close() is implemented by net.UDPConn.  When Close() is called,
+// no new connections are allowed and existing connections are broken.
+// TODO: We could lame-duck this ourselves, if it becomes important.
+type udpUnidlerSocket struct {
+	udpProxySocket
+	eventRecorder record.EventRecorder
+}
+
+func (udp *udpUnidlerSocket) readFromSock(buffer []byte, myInfo *serviceInfo) bool {
+	if !myInfo.isAlive() {
+		// The service port was closed or replaced.
+		return false
+	}
+
+	// Block until data arrives.
+	// TODO: Accumulate a histogram of n or something, to fine tune the buffer size.
+	_, _, err := udp.ReadFrom(buffer)
+	if err != nil {
+		if e, ok := err.(net.Error); ok {
+			if e.Temporary() {
+				glog.V(1).Infof("ReadFrom had a temporary failure: %v", err)
+				return true
+			}
+		}
+		glog.Errorf("ReadFrom failed, exiting ProxyLoop: %v", err)
```

Comment:
I've added some doc to this function, but basically returning false from this function means "don't try to read again from this, and exit the proxy loop instead".